tchannel/tornado/peer.py

```diff
@@ -735,19 +737,11 @@ def choose(self, hostport=None, blacklist=None):
         if hostport:
             return self.get(hostport)
 
-        pop_peers = deque()
-        choose_peer = None
-        n = self.peer_heap.size()
-
-        while n > 0:
-            peer = self.peer_heap.pop_peer()
-            pop_peers.append(peer)
-            if peer.hostport not in blacklist:
-                choose_peer = peer
-                break
-            n -= 1
-
-        for p in reversed(pop_peers):
-            self.peer_heap.push_peer(p)
-
-        return choose_peer
+        try:
+            return self.peer_heap.peek(
+                smallest(
+                    self.peer_heap, (lambda p: p.hostport not in blacklist),
+                ),
+            )
+        except NoMatchError:
+            return None
```

Comment:
To clarify: Up until now, this class depended on only 'self.peer_heap` and its operations. Putting this logic here skips a level from this class' point of view. You're accessing `self.peer_heap` implementation details. You should add a `smallest` method to `PeerHeap` and call that from here.